letsencrypt-auto-source/rebuild_dependencies.py

```diff
@@ -0,0 +1,221 @@
+
+        with open(authoritative_requirements, 'w') as file_h:
+            file_h.write('\n'.join(['{0}=={1}'.format(package, version)
+                                    for package, version in AUTHORITATIVE_CONSTRAINTS.items()]))
+
+        command = ['docker', 'run', '--rm', '--cidfile', cid_file,
+                   '-v', '{0}:/tmp/certbot'.format(CERTBOT_REPO_PATH),
+                   '-v', '{0}:/tmp/workspace'.format(workspace),
+                   '-v', '{0}:/tmp/requirements.txt'.format(authoritative_requirements),
+                   distribution, '/tmp/workspace/script.sh']
+        sub_stdout = sys.stdout if verbose else subprocess.PIPE
+        sub_stderr = sys.stderr if verbose else subprocess.STDOUT
+        process = subprocess.Popen(command, stdout=sub_stdout, stderr=sub_stderr, universal_newlines=True)
+        stdoutdata, _ = process.communicate()
+
+        if process.returncode:
+            if stdoutdata:
+                sys.stderr.write('Output was:\n{0}'.format(stdoutdata))
+            raise RuntimeError('Error while gathering dependencies for {0}.'.format(distribution))
+
+        with open(os.path.join(workspace, 'results'), 'r') as file_h:
+            return file_h.read()
+    finally:
+        if os.path.isfile(cid_file):
+            with open(cid_file, 'r') as file_h:
+                cid = file_h.read()
+            subprocess.call(['docker', 'kill', cid], stdout=None, stderr=None)
+        shutil.rmtree(workspace)
+
+
+def _insert_results(dependencies_map, results, distribution):
+    refined_results = []
+    for result in results.split(os.linesep):
+        match = re.match(r'(.*)==(.*)', result)
+        if match:
+            package = match.group(1)
+            version = match.group(2)
+            if not any(dep in package for dep in ['acme', 'certbot', 'pkg-resources']):
+                dependencies_map.setdefault(package, []).append((version, distribution))
+
+    return refined_results
+
+
+def _process_dependency_map(dependency_map):
+    print('===> Processing the dependency map.')
+    requirements = []
+    conflicts = []
+    for package, versions in dependency_map.items():
+        reduced_versions = _reduce_versions(versions)
+
+        if len(reduced_versions) > 1:
+            version_list = ['{0} ({1})'.format(version, ','.join(distributions))
+                            for version, distributions in reduced_versions.items()]
+            conflict = ('package {0} is declared with several versions: {1}'
+                        .format(package, ', '.join(version_list)))
+            conflicts.append(conflict)
+            sys.stderr.write('ERROR: {0}\n'.format(conflict))
+        else:
+            requirements.append((package, list(reduced_versions)[0]))
+
+    requirements.sort(key=lambda x: x[0])
+    return requirements, conflicts
+
+
+def _reduce_versions(versions):
+    version_map = {}
+    for version in versions:
+        version_map.setdefault(version[0], []).append(version[1])
```

Comment:
There's just too much anonymity here for me to keep track of. Can we introduce some intermediate names, like `for version, distro in versions:`? Also, a docstring would help me figure out the purpose of this function.

Reply:
Done.